### PR TITLE
Replace boost::filysytem by std::filesystem

### PR DESCRIPTION
**Pull request by toly-kournik**
Since std::filesystem became a standard, we want to replace all existing boost::filesystem usages by std::filesystem.
This will also allow not to build boost::filesystem library

### CODE REVIEW

**Comment by yuliasherman** [comment]
Where is the description?

**Comment by toly-kournik** [comment]
The description is in the title

**Comment by yuliasherman** [comment]
Not good enough. What is the reason for such a replacement?

File: util/include/util/filesystem.hpp [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
```
Comment (WildFireFlum) [inline util/include/util/filesystem.hpp:3]: 2022?

File: bftclient/test/CMakeLists.txt [hunk 2]
```diff
@@ -9,6 +9,7 @@ target_link_libraries(bft_client_test PRIVATE
   bftclient_new
   threshsign
   secretsmanager
+  stdc++fs
```
Comment (WildFireFlum) [inline bftclient/test/CMakeLists.txt:12]: https://releases.llvm.org/9.0.0/projects/libcxx/docs/UsingLibcxx.html#using-filesystem

"Starting with LLVM 9.0, support for <filesystem> is provided in the main library and nothing special is required to use <filesystem>."
Why is this needed?

File: util/include/util/filesystem.hpp [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the sub-component's license, as noted in the LICENSE
+// file.
+
+#if __has_include(<filesystem>)
+#include <filesystem>
+namespace fs = std::filesystem;
```
Comment (WildFireFlum) [inline util/include/util/filesystem.hpp:16]: Two questions:
1. Do we use a compiler on which the filesystem library is still experimental?
2. Shouldn't the specific files who want this alias do it? A global two character namespace seems risky to me.

File: bftclient/test/CMakeLists.txt (hunk 2 above)
Comment (toly-kournik) [reply bftclient/test/CMakeLists.txt:12]: what about gcc?

File: util/include/util/filesystem.hpp (hunk 1 above)
Comment (toly-kournik) [reply util/include/util/filesystem.hpp:3]: no idea

Comment (yuliasherman) [reply util/include/util/filesystem.hpp:3]: Needs an update to 2020-2022

File: util/include/util/filesystem.hpp (hunk 3 above)
Comment (toly-kournik) [reply util/include/util/filesystem.hpp:16]: 1. it doesn't hurt
2. although it's a common namespace alias, we can replace it by a longer one.

File: bftclient/test/CMakeLists.txt (hunk 2 above)
Comment (eyalrund) [reply bftclient/test/CMakeLists.txt:12]: what tests did you do to make sure there is no degradation and everything is still working?

Comment (toly-kournik) [reply bftclient/test/CMakeLists.txt:12]: what kind of degradation? We already use std::filesystem in different modules, also it's not used in a performance critical paths, but rarely, mainly during startup.

Comment (eyalrund) [reply bftclient/test/CMakeLists.txt:12]: I ment functionality degradation. making sure everything is still working.

Comment (WildFireFlum) [reply bftclient/test/CMakeLists.txt:12]: > what about gcc?
From https://gcc.gnu.org/gcc-9/changes.html:

Runtime Library (libstdc++)
Improved support for C++17, including:
The C++17 implementation is no longer experimental.
Parallel algorithms and <execution> (requires [Thread Building Blocks](https://github.com/oneapi-src/oneTBB) 2018 or newer).
<memory_resource>.
Using the types and functions in <filesystem> does not require linking with -lstdc++fs now.

What version of gcc is supposed to be supported?
If this is the case, it should be documented in the header.

File: util/include/util/filesystem.hpp (hunk 3 above)
Comment (WildFireFlum) [reply util/include/util/filesystem.hpp:16]: 1. There might be incompatibilities between experimental and final version, which will result is difficult bugs, so If this is not necessary, I wouldn't do it.
2. bft::fs is probably best if we are using our own header.

File: bftclient/test/CMakeLists.txt (hunk 2 above)
Comment (toly-kournik) [reply bftclient/test/CMakeLists.txt:12]: gcc won't compile without experimental and without -stdc++fs linkage.
right now I'm not changing the existing include/linkage behavior

**Review by WildFireFlum — APPROVED** [review]
